### DISCUSSION TITLE
Modules refactor

### DISCUSSION THREAD

**Discussion by Fail-Safe** [sqm-autorate/sqm-autorate#77] [Operational Development]
@Lochnair Hopefully we can move the modules refactor discussion over here to help increase awareness into the work and improvements it will bring to the table.

**Comment by Lochnair** [comment]
Absolutely :)

@chas-iot brought up some good points on the forum I'd like to discuss here later:

> Another piece is a clear definition of the 'contract' for each replaceable component.
> 
> - What are the wakeup criteria?
> - What is the definition of the input or input stream?
> - How can the input stream be extended for additional data items?
> - -  I'm thinking of, for example, Moeller0's request for the actual bytes transferred rather than a relative load).
> - As above for the output.

**Reply by dlakelan** [reply]
I finally had a chance to check out and look at the refactor to modules. On the one hand, I think this is probably a great thing. On the other hand, it seems like a bunch of work and would be easy to have features pushed to the original line of code lost in the refactor... so I'm looking forward to us getting to a "hey people can you test this" version that doesn't have obvious problems (maybe 0.4.1 with stuff we checked in today) and then freeze and look carefully at getting everything into the refactor.

When it comes to the refactor I agree that suddenly "interfaces" become a serious issue. Probably different rate controllers will want different information, and as such a different "baseliner" would be required, which keeps different statistics. Also, the procedure for reselecting new reflectors may not be compatible with certain assumptions in the rate controller, etc. I do think it's worth modularizing, but I don't know that it's going to make it a lot easier for people to come in and do major changes on rate controller algorithms. You're still going to need to understand the scaffolding in multiple places, to know what is currently being done and what/where to make changes to that for your new algorithm.

It might be useful to write up a short (1 page) text document that describes the current "contracts" assumed by the refactor. For example at a high level:

ratecontroller.lua defines a module (table) with two entries "configure" which takes a settings object and returns ... and "ratecontrol" which is called with no arguments and should run a loop that continuously.... reading such and such data from such and such linda object etc.

**Reply by dlakelan** [reply]
Also, I think in the refactor to modules it'd be smart to minimize the number of global variables. Things that are basically static values can remain globals I guess, but anything that could be modified should probably be either an argument to a function or an object communicated between threads through a linda interface.

**Reply by Fail-Safe** [reply]
👍 @dlakelan fully agreed.

**Comment by Fail-Safe** [comment]
Initial refactor to modules has been completed from the "stable" code (`main` branch) into a new testing branch: https://github.com/sqm-autorate/sqm-autorate/tree/testing/refactor-to-modules

Preliminary runtime bugs/kinks have been combed out and fixed. But it is NOT to be considered stable. Test it at your own risk. 😉

That said, for Lua savvy folks, I certainly welcome testing and test feedback. The code still needs several additional passes of cleanup and optimization.

cc: @Lochnair, @dlakelan, @chas-iot

**Reply by Fail-Safe** [reply]
Many thanks to the prior roads paved by @Lochnair and @chas-iot to get the project positioned for modularity. It was no small feat!
